Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp

```diff
@@ -5943,19 +5946,33 @@ IGNORE_CLANG_WARNINGS_END
     {
         JSGlobalObject* globalObject = m_graph.globalObjectFor(m_origin.semantic);
         ArrayMode arrayMode = m_node->arrayMode();
+
+        // This function is used by `ArrayAt` but it supports only Int32, Contiguous, Double.
+        ASSERT(m_node->op() != ArrayAt || (arrayMode.type() == Array::Int32 || arrayMode.type() == Array::Contiguous || arrayMode.type() == Array::Double));
+
         switch (arrayMode.type()) {
         case Array::Int32:
         case Array::Contiguous: {
-            LValue index = lowInt32(m_graph.varArgChild(m_node, 1));
+            LValue originalIndex = lowInt32(m_graph.varArgChild(m_node, 1));
             LValue storage = lowStorage(m_graph.varArgChild(m_node, 2));
 
+            LValue arrayLength = m_out.load32NonNegative(storage, m_heaps.Butterfly_publicLength);
+            LValue index = m_node->op() == ArrayAt
+                ? m_out.select(m_out.lessThan(originalIndex, m_out.int32Zero), m_out.add(arrayLength, originalIndex), originalIndex)
+                : originalIndex;
+
             IndexedAbstractHeap& heap = arrayMode.type() == Array::Int32 ?
                 m_heaps.indexedInt32Properties : m_heaps.indexedContiguousProperties;
 
             LValue base = lowCell(m_graph.varArgChild(m_node, 0));
 
             if (arrayMode.isInBounds()) {
-                LValue result = m_out.load64(baseIndexWithProvenValue(heap, storage, index, m_graph.varArgChild(m_node, 1)));
+                if (m_node->op() == ArrayAt)
+                    speculate(OutOfBounds, noValue(), nullptr, m_out.aboveOrEqual(index, arrayLength));
+                LValue result = m_out.load64(
+                    m_node->op() == ArrayAt
+                        ? m_out.baseIndex(heap, storage, m_out.zeroExtPtr(index))
+                        : baseIndexWithProvenValue(heap, storage, index, m_graph.varArgChild(m_node, 1)));
```

Comment:
I think in this case we know we're using originalIndex so we can just use `baseIndexWithProvenValue(heap, storage, originalIndex, m_graph.varArgChild(m_node, 1)));`